### PR TITLE
Unit Tests

### PR DESCRIPTION
**Pull request by SantamRC**
To run :
```
python dev/generate-tests.py
python -m pytest -vv unit-tests
```

### CODE REVIEW

File: dev/generate-tests.py [hunk 1]
```diff
@@ -692,8 +700,46 @@ class Error(ctypes.Structure):
...
+                        file.write("\tassert " + key + " == " + "pytest_" + key + "\n")
+                file.write("\n\n")
+
+
 if __name__ == "__main__":
     specdict = readspec()
     genspectests(specdict)
     gencpukerneltests(specdict)
     gencudakerneltests(specdict)
+    generateTests()
```
Comment (ianna) [inline dev/generate-tests.py]: @SantamRC - this should probably follow previous function names convention: `genunittests`?

File: test-data.yml [hunk 2]
```diff
@@ -0,0 +1,2326 @@
...
+        starts_out: [0, 1, 1, 1, 1, 1, 1, 1]
+        stops_out: [1, 1, 1, 1, 1, 1, 1, 1]
+  - name: awkward_content_reduce_zeroparents_64
+    tests:
+    - error: true
+      inputs: {}
+      outputs:
+        toparents: [0]
+    - error: false
+      inputs: {length: 10}
```
Comment (ianna) [inline test-data.yml]: btw, why does this input look different?

File: test-data.yml [hunk 3]
```diff
@@ -0,0 +1,2326 @@
...
+        length: 8
+        offsets_in: [0, 1, 1, 1, 1, 1, 1, 1]
+      outputs:
+        mask_out: [0, 1, 2, 3, 4, 5, 6, 7]
+        starts_out: [0, 1, 1, 1, 1, 1, 1, 1]
+        stops_out: [1, 1, 1, 1, 1, 1, 1, 1]
+  - name: awkward_content_reduce_zeroparents_64
+    tests:
+    - error: true
+      inputs: {}
```
Comment (ianna) [inline test-data.yml]: I think, `length` in this case is `0` and the test should pass.

**Review by ianna — COMMENTED** [review]
It looks great, thanks! I've added some minor comments.

Also from pre-commit:
`dev/generate-tests.py:728:34: E712 comparison to True should be 'if cond is True:' or 'if cond:'`

File: test-data.yml (hunk 2 above)
Comment (SantamRC) [reply test-data.yml]: The format is generated like this. Looking into it.

File: test-data.yml (hunk 3 above)
Comment (SantamRC) [reply test-data.yml]: The definition is,
```
def awkward_content_reduce_zeroparents_64(toparents, length):
    for i in range(length):
        toparents[i] = 0
```
So `length=0` will not change `toparents` but if `length=1` then `toparents=[0]` and the test will pass. Should I put the length as 0 or 1 or should I keep it as an error case?

**Comment by jpivarski** [comment]
@all-contributors please add @SantamRC  for test

**Pull request merged**